### PR TITLE
Reuse of token in TOTP is possible

### LINKED ISSUE #13607: Reuse of token in TOTP is possible
### Describe the bug

According to the document, the verifier should not accept the second attempt of the OTP after successful validation but KeyCloak is allowing using of the same OTP token multiple times within the timeframe.  

https://www.rfc-editor.org/rfc/rfc6238.html#section-5

Note that a prover may send the same OTP inside a given time-step window
multiple times to a verifier. The verifier MUST NOT accept the second
attempt of the OTP after the successful validation has been issued for the
first OTP, which ensures one-time only use of an OTP.

### Version

18.0.0

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Fixes #13607

PR to the documentation: https://github.com/keycloak/keycloak-documentation/pull/1667
PR to the new admin console: https://github.com/keycloak/keycloak-ui/pull/3147

### CODE REVIEW

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties
Comment (stianst) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties:254]: Should most likely not leak that this was in fact a valid code.

Comment (mabartos) [reply themes/src/main/resources/theme/base/login/messages/messages_en.properties:254]: Ok, make sense. So, only show to the user 'Invalid authenticator code', right?

**Comment by mabartos** [comment]
Added support for the old admin console. I'll also attach PR to the new admin console here and documentation.

I also need to find out, why the `codeReusable` flag is always 'false' as the default value should be 'true'. It's not possible to change the value on the server and store it in the DB, so I need to investigate it a little bit.

File: kubernetes/keycloakrealmimports.keycloak.org-v1.yml
Comment (mabartos) [inline kubernetes/keycloakrealmimports.keycloak.org-v1.yml:1305]: Not sure about this.

**Comment by mabartos** [comment]
Do we need to consider the probability of collision of OTP codes after some time? As we store the last code, after some time, there can be a collision with that (especially when the number of required digits is low). 

We could delete the last value after some time (f.e. after the 10th OTP period).

It's just a corner case, but it could happen. Or just keep the user waiting for the next OTP period?

File: server-spi/src/main/java/org/keycloak/models/OTPPolicy.java
Comment (mabartos) [inline server-spi/src/main/java/org/keycloak/models/OTPPolicy.java:77]: Will be the new behavior the default one, or must the admin explicitly define that those OTPs cannot be reused?

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:117]: I would remove this log entry as it is not shown to the user. If we want to enable logging, we should also be adding to other places too like when the OTP is just invalid.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/OtpPolicyTest.java
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/OtpPolicyTest.java:37]: Tests are always a good thing to include. But I'm not sure if we need one for this purpose because we are indirectly checking if values are properly being set in other tests.

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:121]: I think it should be safer to validate codes based on the last successful validation interval rather than the last used code.  Although not a big issue when it happens, it should be possible that the same code is generated for different times. Codes are 6 or 8 digits and I think it should happen.

By doing checks based on the interval, we store the last interval from here https://github.com/keycloak/keycloak/blob/fd2787ae7da15e50a523d52c59617584e6d09cc7/server-spi-private/src/main/java/org/keycloak/models/utils/TimeBasedOTP.java#L83.

Then we change the `validateTOTP` method to accept the last validation interval and introduce a check like:

```
if (currentInterval <= lastValidationInterval) {
    return false;
}
```

**Comment by pedroigor** [comment]
@mabartos Very nice work. Thanks for looking at this. Just a few comments, most of them nitpicks.

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (mabartos) [reply services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:121]: @pedroigor Thanks a lot for your suggestion! I'm aware that the solution with the last used code is not so elegant and brings other problems to consider.

However, I've been thinking a little bit about it, and the approach, which you've suggested, has some drawbacks as well. The thing is related to the "look around window". The code can be valid even for the following auth windows (the count of valid windows depends on the `lookAroundWindow` field).  I assume that you had in mind, for the successful validation, store either:

1) `lastValidationInterval = currentInterval;` -> Store the last "success" interval and then use the condition, that you've mentioned above. It'd work only without the "look around window" feature. A user can also enter the code from the previous auth window. Moreover, as the code is valid even for the next window, an attacker could use the same code in the next window.

or

2) `lastValidationInterval = currentInterval + lookAroundWindow;` ->  Store the last "success" interval + look around window value (or the evaluation of the valid interval can be done later). It'd mean, that the user is not able to use the code from the next auth window and needs to wait till `currentInterval > lastValidationInterval` - not okay.

--------------------------------------------------------------------
I think the combination of both could work; store the last successful OTP code and the last successful interval. I can assemble a particular solution for that and you can review it later.

If I should elaborate it a little bit more, just feel free to let me know :))

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (mabartos) [reply services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:117]: Ok, agree.

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (mabartos) [reply services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:115]: Ok, agree. Thanks!

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmEntity.java
Comment (mabartos) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmEntity.java:99]: +1

File: server-spi/src/main/java/org/keycloak/models/OTPPolicy.java
Comment (mabartos) [reply server-spi/src/main/java/org/keycloak/models/OTPPolicy.java:77]: Agree :)

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (pedroigor) [reply services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:121]: Yeah, we need to consider the look ahead window and for that, we should take into account the `adjust` interval when storing the last successful interval.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/OtpPolicyTest.java
Comment (mabartos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/policy/OtpPolicyTest.java:37]: Yep, you're right... The OTP policy setting is already included in other tests.

File: kubernetes/keycloakrealmimports.keycloak.org-v1.yml
Comment (pedroigor) [reply kubernetes/keycloakrealmimports.keycloak.org-v1.yml:1305]: TBH, have no clue. Perhaps we can remove for now and update this code later depending on what @vmuzikar has to say.

**Comment by pedroigor** [comment]
@mposolda This PR should now introduce a write operation to storage every time the users successfully authenticate using TOTP. We could potentially store the information in memory but:

* It means eventual security because the last successful interval will be lost during cluster restarts
* The new store does not have any cache layer

Wdyt?

**Comment by mabartos** [comment]
@pedroigor IMHO, the interval shouldn't be stored in memory due to those things you've mentioned. It'd be possible to (most of the time) reuse OTP codes of the particular credential(TOTP) in the "cluster" environment, where more than one instance is used with some load balancer - we do not want this. It should be shared among all instances.

**Comment by mposolda** [comment]
@pedroigor @mabartos I believe that for this type of use-cases, we can use `SingleUseTokenStoreProvider` ? Probably method `putIfAbsent` on that provider. That should guarantee the cluster-safe and crossdc-safe and is supported also by both old and new store. As a key, we can use something like `<user-uuid>.<totp-code>` (The `totp-code` alone is not sufficient IMO as it is possible that 2 users are trying to login at the same time with their OTP and for both of them, there is same code like `123456`).

WDYT?

**Comment by mabartos** [comment]
@mposolda I'm not sure how exactly the `SingleUseTokenStoreProvider` works, but it could be a promising solution for that in order to avoid accessing directly DB(store). 

At this moment, we store the last successful interval and I'd say the more suitable key for that should be more credential-related as the credential is already associated with the user and unique. Something like `<credential-id>.<last-valid-interval>`. WDYT?

**Comment by mposolda** [comment]
@mabartos Yes, I think that `<credential-id>.<last-valid-interval>` is likely even better.

The `SingleUseTokenStoreProvider` is exactly for the similar use-cases and it is already implemented in the stores. For example it is used in OAuth2 Device Flow for check the polling interval and/or for checking that client authentication with JWT doesn't reuse same JWT multiple times. I suggest to search for `SingleUseTokenStoreProvider.putIfAbsent` and you will see :-)

One note is to use the correct timeout - I guess it is possible to "compute" it based on the settings of Totp policy (it probably depends on the period and lookAheadWindow)

**Comment by mabartos** [comment]
@mposolda Thanks for your advice! I assume that you probably mean `SingleUseObjectProvider` right? 

I'll try to integrate it here tomorrow.

**Comment by mposolda** [comment]
@mabartos Yes, I am sorry for the confusion. I think the `SingleUseTokenStoreProvider` was the original name, but it was renamed at some point :-)

**Comment by pedroigor** [comment]
@mabartos If we are going to keep it persisted, I would just keep it as is and store it in the database. As @mabartos mentioned, it is just another metadata associated with the credential and as such should be close as possible to it.

Not hard on it though if you prefer otherwise.

**Comment by pedroigor** [comment]
> I might be mistaken in this thought, but I'd go with the DB approach as well. I'm aware that's not an ideal solution to store the whole credential entity after successful verification of the OTP, but I'd say it's the only solution for now. Perhaps, it'd be good to have the possibility to save the value only as some attribute in a map structure, where wouldn't be necessary to update the whole entity, but only some particular record. Does it make sense?

For me it does. I would say that we should rely on how the new store and the optimizations it does when persisting data. I think we should also keep credential-related data close to each other. For last, using a single object provider is not going to persist data when doing legacy store (we still need to support it) and introduce different behavior between legacy and new store when validating token reuse.

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:119]: I see one small issue that this does not strictly guarantee that code is used really only once. If there are 2 concurrent callers trying to verify code `123456` it can happen that:
- thread1 (EG. valid user) calls this validation and successfully manage to verify `123456`
- thread2 (EG. potential attacker, who somehow stole the last code of valid user) calls this validation and successfully manage to verify `123456`
- thread1 updates DB. However there was already successful validation of the attacker from thread2 (FAIL)

Not sure how realistic this approach is, but just pointing it :-)

**Review by mposolda — COMMENTED** [review]
@mabartos @pedroigor I am sorry as I meant to use `<credential-id>.<totp-code>` as a key for the `SingleUseObjectProvider` where the `totp-code` is something user provided (EG. `123456`). I did not meant to use `<credential-id>.<last-valid-interval>`. I did not looked at the code before and I missed what `last-valid-interval` is :-)

And I believe the period to use when calling `SingleUseObjectProvider.putIfAbsent` can be something like `totpPeriod * (lookAheadWindow + 1) * 2`. So for example 90 in case that configured period in the policy is 30 and configured lookAheadWindow is 1.
This should guarantee that user can authenticate only once with the code `123456` for the given period.

With this approach, there is no need to change anything in `TimeBasedOTP` class and credential format and no need to deal with `lastValidInterval` at all. Also `SingleUseObjectProvider.putIfAbsent` is guaranteed to work with both old and new store.

Your approach with `lastValidinterval` works as well, however I see these potential problems:
- There is a need to update DB during each authentication as you pointed. It is likely more expensive than updating infinispan or JDG server (in the case of old store).
- I think it is possible that 2 concurrent callers would be able to successfully authenticate with same code (This should not be possible with the `SingleUseObjectProvider` approach as it guarantee `putIfAbsent` semantics across all servers in the cluster or cross-dc environment). See my inline comment.

But I also don't strongly insist on this if both of you guys believe otherwise :-)

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (mabartos) [reply services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java]: Interesting thought! Thanks for that. ATM, the new solution doesn't update DB.

File: kubernetes/keycloakrealmimports.keycloak.org-v1.yml
Comment (mabartos) [reply kubernetes/keycloakrealmimports.keycloak.org-v1.yml]: Ok

**Comment by mabartos** [comment]
@mposolda @pedroigor Sorry for the confusion. I've just squashed those commits (started to be a little messy) and provided a new approach for that. 

I think this solution might be promising as there's no update to the Credential entities and there's used `SingleUseObjectProvider`(SUO) for the last code. 

The lifespan of the OTP code, stored by the SUO provider, is set to the highest one regarding the used "look around window". I don't think it's necessary to make some optimizations there as the probability of duplicate code is almost zero for the short time. 

I also think that we don't need to provide this functionality as was present with the `lastValidInterval` -- when you use some OTP code, you're not able to use previous code for that within the "look around window". It should be possible to accept even the previous code as the credential is related to the user, not the device. I mean, the user can use multiple devices and the clock synchronization might be different for those devices and the server itself.

**Comment by pedroigor** [comment]
But now you have legacy allowing core reuse for the legacy store if the cluster restarts. Differently from the new store where the code is persistent.

Also, AFAIK, the new store will always update the dabatase (if using map-jpa) so I don't think this is an issue as lot of optimizations are being done to make it possible to run the server without any cache layer but the database.

**Comment by pedroigor** [comment]
> * I think it is possible that 2 concurrent callers would be able to successfully authenticate with same code (This should not be possible with the `SingleUseObjectProvider` approach as it guarantee `putIfAbsent` semantics across all servers in the cluster or cross-dc environment). See my inline comment.

Do you mean an attacker was able to steal the user account completly? If so, code reuse will make no difference as the attacker can try any time window.

**Comment by mposolda** [comment]
> But now you have legacy allowing core reuse for the legacy store if the cluster restarts. Differently from the new store where the code is persistent.

Yes, true. But I am not sure how big this issue really is in the reality? Considering that cluster restart (or server restart in case of the singlenode environment) takes some time and TOTP code is valid just for the small period of time. In most cases the TOTP code would be probably invalid before the server boots up. But agree that this corner case is slight disadvantage of the `SingleUseObjectProvider` approach.

In general, it is the price to pay for using infinispan instead of the DB (infinispan has better performance, but doesn't survive server restart). We already have this corner case for other things relying on the single use (EG. JWT used in client authentication , probably also OAuth2 `code` parameter). The possibility is to have `actionTokens` cache to be backed by the persistent store, but TBH I am not sure if this worth an effort at this point?

> Also, AFAIK, the new store will always update the dabatase (if using map-jpa) so I don't think this is an issue as lot of optimizations are being done to make it possible to run the server without any cache layer but the database.

Yes, on the other hand new store can be changed in the future and somehow optimized, so the `SingleUseObjectProvider.putIfAbsent` won't use the (expensive) DB in the future.

**Comment by mposolda** [comment]
> > * I think it is possible that 2 concurrent callers would be able to successfully authenticate with same code (This should not be possible with the `SingleUseObjectProvider` approach as it guarantee `putIfAbsent` semantics across all servers in the cluster or cross-dc environment). See my inline comment.
> 
> Do you mean an attacker is able to steal the user account completely? If so, code reuse will make no difference as the attacker can try any time window.

I meant more a scenario where attacker is able to intercept HTTP request somehow and see that the HTTP request sent by the user uses `totp=123456` . Attacker will be then able to replicate this request right after. So kind of the "replay attack" .

**Comment by mposolda** [comment]
> I also think that we don't need to provide this functionality as was present with the `lastValidInterval` -- when you use some OTP code, you're not able to use previous code for that within the "look around window". It should be possible to accept even the previous code as the credential is related to the user, not the device. I mean, the user can use multiple devices and the clock synchronization might be different for those devices and the server itself.

Yes, that is also corner case, but IMO it shows another advantage of the `putIfAbsent` approach. In fact, specification mentions that code must be single-use, which means to me that `putIfAbsent` approach is more aligned with the specs then the `lastInterval` approach.

File: services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java
Comment (mabartos) [inline services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java:115]: Note: The validation of OTP should be executed prior to this.

Comment (mabartos) [reply services/src/main/java/org/keycloak/credential/OTPCredentialProvider.java]: Fixed

**Comment by pedroigor** [comment]
@mabartos Works for me.

**Comment by pedroigor** [comment]
> In general, it is the price to pay for using infinispan instead of the DB (infinispan has better performance, but doesn't survive server restart). We already have this corner case for other things relying on the single use (EG. JWT used in client authentication , probably also OAuth2 `code` parameter). The possibility is to have `actionTokens` cache to be backed by the persistent store, but TBH I am not sure if this worth an effort at this point?

I'm not really trying to compare in-memory vs persistent storage. My point here is the difference in *behavior* once you migrate to the new store in the future. IIRC, the new store is focusing on making it possible to run the server without anything else but a database. If you are migrating from legacy to the new store you now have the same behavior as if you were storing the last time interval in the legacy store. And the same issues you are pointing out: potential degradation in performance, possibility to reuse the code when making concurrent requests to different servers.

To put more fire into the discussion :) You are now saying that we should prefer doing in-memory but the new store is not about forcing users to deploy/manage an infinispan server but using it for specific use cases like a cross data center. For the majority of the use cases, a single database should be enough.

That said, we don't quite need to discuss about `actionTokens` because the code using it in legacy is already storing data to the database if using the new store already :) 

> > Also, AFAIK, the new store will always update the dabatase (if using map-jpa) so I don't think this is an issue as lot of optimizations are being done to make it possible to run the server without any cache layer but the database.

> Yes, on the other hand new store can be changed in the future and somehow optimized, so the `SingleUseObjectProvider.putIfAbsent` won't use the (expensive) DB in the future.

It should be possible already to use a store other than database on a per "storage area" basis. But again, the new store goal (AFAIK) is not really force users to deploy hot rod whenever they want more performance. There is a huge cost here.

**Comment by pedroigor** [comment]
> I meant more a scenario where attacker is able to intercept HTTP request somehow and see that the HTTP request sent by the user uses `totp=123456` . Attacker will be then able to replicate this request right after. So kind of the "replay attack" .

Again if a MITM is happening worse things can be done :) Like steal the account completely. I would also consider that OIDC/OAuth2 relies on TLS to avoid this and many other forms of attacks.

**Comment by mposolda** [comment]
> > In general, it is the price to pay for using infinispan instead of the DB (infinispan has better performance, but doesn't survive server restart). We already have this corner case for other things relying on the single use (EG. JWT used in client authentication , probably also OAuth2 `code` parameter). The possibility is to have `actionTokens` cache to be backed by the persistent store, but TBH I am not sure if this worth an effort at this point?
> 
> I'm not really trying to compare in-memory vs persistent storage. My point here is the difference in _behavior_ once you migrate to the new store in the future. IIRC, the new store is focusing on making it possible to run the server without anything else but a database. If you are migrating from legacy to the new store you now have the same behavior as if you were storing the last time interval in the legacy store (probably better performance because the new store has a lot of optimizations). And the same issues you are pointing out: potential degradation in performance.
> 
> To put more fire into the discussion :) You are now saying that we should prefer doing in-memory but the new store is not about forcing users to deploy/manage an infinispan server but using it for specific use cases like for cross data center. For the majority of the use cases, a single database should be enough.

Oh, sorry. I did not want to say that in-memory is always a best. But maybe it could be implied from previous post, I am sorry for the confusion then.

What I mean in nutshell is, that we have API method `SingleUseObjectProvider.putIfAbsent`, which is exactly useful for this purpose of storing the single-use things and making the sure that they are really single-use. I see it in a way that it is up to the store how exactly this could be implemented. In the old store, it was done in infinispan cache (in case of singlenode or cluster) or in JDG server (in case of cross-dc). In the new store, it is in the DB, but I can see that it may be also in the future potentially in hotrod or somewhere else for some deployments.

IMO it would be nice if this use-case `make sure that something is single-use for some small period of time` is handled in a unified way for all the similar use-cases (EG. single-use TOTP codes, single-use JWT used in JWT client authentication, single-use OAuth2 code parameter etc) and there is no need to have custom solution for every kind of the similar use-case.

When you mentioned: `My point here is the difference in _behavior_ once you migrate to the new store in the future.` . Did you mean that after migration , the TOTP codes, which were stored in the infinispan/memory won't be available in the DB after migration? This is true, but I don't consider it as an important issue. The TOTP codes are short-lived and `SingleUseObjectProvider.putIfAbsent` is meant for the short-lived things (note that it has the `lifespanInSeconds` parameter). In other words this scenario:
- Old store is used. User logins with TOTP code `123456`
- Now server is restarted and migrated to the new store
- User/attacker may try to authenticate again with code `123456` . Now re-using the code `123456` would be possible due it was not migrated from previous version. However the TOTP code `123456` very likely not be valid anymore due the restart + migration probably takes few minutes at least.

This problem exists again for the other single-use things (OAuth2 code parameters and others). However due their short-lived nature and due the fact, that migration to the new store will take some time, I don't think that it is a big issue that those are not available after restart. Or do you think otherwise?

I consider bigger issue that all users need to re-login as user sessions will be also lost after migration (TBH I am not very familiar with the new store and not sure what data exactly we would provide to migrate).

> 
> That said, we don't quite need to discuss about `actionTokens` because the code "using" it in legacy is already storing data to the database if using the new store already :)
> 
> > > Also, AFAIK, the new store will always update the dabatase (if using map-jpa) so I don't think this is an issue as lot of optimizations are being done to make it possible to run the server without any cache layer but the database.
> 
> > Yes, on the other hand new store can be changed in the future and somehow optimized, so the `SingleUseObjectProvider.putIfAbsent` won't use the (expensive) DB in the future.
> 
> It should be possible already to use a store other than database on a per "storage area" basis. But again, the new store goal (AFAIK) is not really force users to deploy hot rod whenever they want more performance. There is a huge cost here.

Yes, I agree. I think that DB works fine if there is nothing more optimal available. I mean that when you call `SingleUseObjectProvider.putIfAbsent`, you would simply rely on the store implementation to provide the most optimal implementation for the particular deployment.

**Comment by mposolda** [comment]
> > I meant more a scenario where attacker is able to intercept HTTP request somehow and see that the HTTP request sent by the user uses `totp=123456` . Attacker will be then able to replicate this request right after. So kind of the "replay attack" .
> 
> Again, if a MITM is happening worse things can be done :) Like steal the account completely. I would also consider that OIDC/OAuth2 relies on TLS to avoid this and many other forms of attacks.

Completely agree that OIDC relies on TLS. On the other hand, it provides some more "protection layers" to not rely solely on the TLS. I see that for example single-use OAuth2 code is one of those additional protections. Similarly the single-use TOTP codes. Like if we rely just on the TLS, then we don't need to do at all this exercise with the single-use TOTP codes? You know that TLS prevent somethings, but cannot prevent every possibility how the TOTP code can be "stolen" and replayed by the attacker (for example eavesdropping is another possible scenario)

> Not sure if this is a strong argument to favor putIfAbsent.

I don't consider it as very strong argument, more as a corner case. I think that approach with `lastInterval` would work as well. At the same time, I don't see any strong argument for the `lastInerval` approach though (I consider the fact it doesn't survive cluster restart and migration to the new store also as a corner case... Do you think otherwise?)

Overally, I would personally prefer to rely on the unified solution for all the similar use-cases  (which IMO `SingleUseObjectProvider.putIfAbsent` is supposed to provide) rather than inventing the wheel for anything similar in the future.

**Comment by pedroigor** [comment]
@mposolda @mabartos Works for me.

**Comment by mposolda** [comment]
@pedroigor @mabartos Thanks for the discussion around this!

**Comment by mabartos** [comment]
@mposolda @pedroigor Yep, thanks! It's great to see a lot of thoughts and ideas from different points of view. Really valuable and inspiring to me! :)))

Btw. I'll try to fix those tests next week.

**Comment by hmlnarik** [comment]
This code affects storage in non-backwards compatible way. The storage team needs review and potential updates of this work before this could be merged.

**Review by vramik — COMMENTED** [review]
Thanks @mabartos. Test failures in GHA are relevant, the cause for it is that there is realm attribute used in legacy store for storing REALM_REUSABLE_CODE_ATTRIBUTE while in map store it's stored as class field. I'd like to understand the reasoning behind that decision. Thank you.

@hmlnarik I've checked the migration of jpa map storage and I believe there is no issue in current state. Newly added filed is in the end stored within json and it is not searchable field, nor generated field.

**Comment by mabartos** [comment]
@vramik Thanks for your comment and your point about the different stores used for that. 

The reason, why the realm property `isOtpPolicyCodeReusable` is stored in the legacy store as an attribute is, that we don't wanna change the DB schema just because of this one property (and potentially provide a migration script?).

And for the Map-store, IMHO, it's better to have this field visible and not hidden in attributes as there are no other required steps for integrating it there; no drawbacks (AFAIK).

However, you have much more experience with those stores, so what would you suggest?

**Comment by mabartos** [comment]
The RealmTest should be fixed now.

**Comment by mabartos** [comment]
@mposolda @pedroigor Guys, do you have a chance to take a look at it? Thank you very much.

**Review by vramik — COMMENTED** [review]
Thanks @mabartos. It makes sense to me. As stated earlier from map-jpa point of view migration is not needed, therefore LGTM. Let's wait for @mhajas to see how it's in hot-rod.

**Review by pedroigor — APPROVED** [review]

**Review by mposolda — APPROVED** [review]
@mabartos Thanks, I am approving. But I did not review parts related to the new store - so this can be merged only once it is approved by members of the store team as discussed above.

NOTE: This PR is likely going to be shown during the F2F as the "Demo PR", which requires collaboration from the multiple teams (storage, auth, new admin console, docs) and there might be discussion on how we handle the similar PRs in the future etc. Just for this reason, it may be good to keep it open for now. So hope you don't mind if this PR is merged likely after f2f :-)

**Comment by mabartos** [comment]
@pedroigor @mposolda Thanks for your review!

Yep, I don't mind merging this PR after the F2F. We need to wait for the @mhajas's approval anyway, and AFAIK, he has PTO for some time ATM. 

On the other hand, I don't know if it's a good approach to wait another, at least, 2 weeks for merging this fix. To be more specific, this fix should be included in the next micro release of RH-SSO 7.6.x as well, right?

**Pull request merged**
